src/services/fetchers/UniswapV3MultiFetcher.ts

```diff
@@ -0,0 +1,101 @@
+import {StaticJsonRpcProvider} from '@ethersproject/providers';
+import {inject, injectable} from 'inversify';
+import {Logger} from 'winston';
+
+import {PoolRepository} from '../../repositories/PoolRepository.js';
+import {DexProtocolName} from '../../types/DexProtocolName.js';
+import {ChainsIds} from '../../types/ChainsIds.js';
+import {ContractHelperRepository} from '../../repositories/ContractHelperRepository.js';
+import {BigNumber} from 'ethers';
+
+export type UniswapV3MultiFetcherParams = {
+  token0: string;
+  token1: string;
+};
+
+export interface OutputValues {
+  token0: string;
+  token1: string;
+  value: number;
+}
+
+@injectable()
+class UniswapV3MultiFetcher {
+  provider: StaticJsonRpcProvider | undefined;
+  logPrefix = '[UniswapV3MultiFetcher]';
+  readonly dexProtocol = DexProtocolName.UNISWAP_V3;
+  readonly chainId = ChainsIds.ETH;
+
+  @inject('Logger') protected logger!: Logger;
+  @inject(PoolRepository) protected poolRepository!: PoolRepository;
+  @inject(ContractHelperRepository) contractRepository!: ContractHelperRepository;
+
+  async apply(inputs: UniswapV3MultiFetcherParams[]): Promise<OutputValues[]> {
+    this.logger.debug(`${this.logPrefix}: start with inputs ${JSON.stringify(inputs)}`);
+
+    const poolsToFetch: {
+      pools: string[];
+      base: string;
+      quote: string;
+    }[] = [];
+
+    for (const input of inputs) {
+      const {token0, token1} = input;
+      this.logPrefix = '[UniswapV3MultiFetcher]';
+
+      const data = await this.poolRepository.find({
+        protocol: this.dexProtocol,
+        token0,
+        token1,
+      });
+
+      if (data.length === 0) {
+        this.logger.error(`${this.logPrefix} no data found for ${token0}-${token1}`);
+        continue;
+      }
+
+      this.logger.debug(`${this.logPrefix} data found ${token0}-${token1}`);
+      const pools = data.map((item) => item.pool);
+      poolsToFetch.push({pools, base: token0, quote: token1});
+    }
+
+    if (poolsToFetch.length === 0) {
+      this.logger.error(`${this.logPrefix} no data to fetch`);
+      return [];
+    }
+
+    const prices = await this.fetchData(poolsToFetch);
+
+    return prices;
+  }
+
+  private async fetchData(poolsToFetch: {pools: string[]; base: string; quote: string}[]): Promise<OutputValues[]> {
+    const contract = this.contractRepository.get(this.dexProtocol).getContract(this.chainId);
+    const [results] = (await contract.callStatic.getPrices(poolsToFetch)) as {success: boolean; price: BigNumber}[][];
+    const outputs: OutputValues[] = [];
```

Comment:
We could move to line 82 since when it is first used.

Reply:
True, I didn't notice it.

Also `return []` makes it easier to see what it returns at that moment without having to go up and see what `outputs` are at that point.

Reply:
Better indeed. Done